[BUG] Java heap space error when trying to run pulsar sink

**Describe the bug**
https://streamnative.io/en/blog/tech/2020-10-20-cloud-storage-sink-connector-251 When running part 2 in step 2, we encountered java heap space error

**To Reproduce**
Steps to reproduce the behavior:
1. copy pulsar-io-cloud-storage-2.5.1.nar to apache-pulsar-2.7.0/connectors
2. restart broker
3. upload cloud-storage-sink-config.yaml to apache-pulsar-2.7.0 folder
4. bin/pulsar-admin sink localrun --sink-config-file cloud-storage-sink-config.yaml
5. see java heap space error: [function-timer-thread-6-1] ERROR org.apache.pulsar.functions.runtime.RuntimeSpawner - demotenant/demonamespace/Cloud-Storage-sink-java.lang.OutOfMemoryError: Java heap space Function Container is dead with exception.. restarting

**Expected behavior**
Should not see error

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Our brokers are deployed in docker containers, so step 4 was run within the docker container.

Can anyone help with this? Thanks

This is because [BlobStoreAbstractConfig.validate()](https://github.com/streamnative/pulsar-io-cloud-storage/blob/master/src/main/java/org/apache/pulsar/io/jcloud/BlobStoreAbstractConfig.java#L89) method allows endpoint property value to be passed in incorrect format which results in exception being caught in [JavaInstanceRunnable](https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L782) and OOM when trying to log it. 

Issue could be reproduced by setting `endpoint: "s3.us-east-1.amazonaws.com"` in the config file.

Complete stack trace for the OOM is: 
`  at java.lang.OutOfMemoryError.<init>()V (OutOfMemoryError.java:48)
  at java.util.Arrays.copyOf([CI)[C (Arrays.java:3332)
  at java.lang.AbstractStringBuilder.ensureCapacityInternal(I)V (AbstractStringBuilder.java:124)
  at java.lang.AbstractStringBuilder.append(Ljava/lang/String;)Ljava/lang/AbstractStringBuilder; (AbstractStringBuilder.java:448)
  at java.lang.StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder; (StringBuilder.java:136)
  at java.lang.StringBuilder.append(Ljava/lang/String;)Ljava/lang/AbstractStringBuilder; (StringBuilder.java:76)
  at java.lang.AbstractStringBuilder.append(Ljava/lang/CharSequence;)Ljava/lang/AbstractStringBuilder; (AbstractStringBuilder.java:484)
  at java.lang.StringBuilder.append(Ljava/lang/CharSequence;)Ljava/lang/StringBuilder; (StringBuilder.java:166)
  at java.lang.StringBuilder.append(Ljava/lang/CharSequence;)Ljava/lang/Appendable; (StringBuilder.java:76)
  at java.util.Formatter$FormatSpecifier.print(Ljava/lang/String;)V (Formatter.java:2913)
  at java.util.Formatter$FormatSpecifier.printString(Ljava/lang/Object;Ljava/util/Locale;)V (Formatter.java:2886)
  at java.util.Formatter$FormatSpecifier.print(Ljava/lang/Object;Ljava/util/Locale;)V (Formatter.java:2763)
  at java.util.Formatter.format(Ljava/util/Locale;Ljava/lang/String;[Ljava/lang/Object;)Ljava/util/Formatter; (Formatter.java:2520)
  at java.util.Formatter.format(Ljava/lang/String;[Ljava/lang/Object;)Ljava/util/Formatter; (Formatter.java:2455)
  at com.google.inject.internal.Errors.format(Ljava/lang/String;Ljava/util/Collection;)Ljava/lang/String; (Errors.java:562)
  at com.google.inject.CreationException.getMessage()Ljava/lang/String; (CreationException.java:48)
  at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(Ljava/lang/Throwable;Ljava/util/Set;)V (ThrowableProxy.java:134)
  at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(Ljava/lang/Throwable;)V (ThrowableProxy.java:122)
  at org.apache.logging.log4j.core.impl.Log4jLogEvent.getThrownProxy()Lorg/apache/logging/log4j/core/impl/ThrowableProxy; (Log4jLogEvent.java:566)
  at org.apache.logging.log4j.core.pattern.ExtendedThrowablePatternConverter.format(Lorg/apache/logging/log4j/core/LogEvent;Ljava/lang/StringBuilder;)V (ExtendedThrowablePatternConverter.java:64)
  at org.apache.logging.log4j.core.pattern.PatternFormatter.format(Lorg/apache/logging/log4j/core/LogEvent;Ljava/lang/StringBuilder;)V (PatternFormatter.java:38)
  at org.apache.logging.log4j.core.layout.PatternLayout$PatternSerializer.toSerializable(Lorg/apache/logging/log4j/core/LogEvent;Ljava/lang/StringBuilder;)Ljava/lang/StringBuilder; (PatternLayout.java:334)
  at org.apache.logging.log4j.core.layout.PatternLayout.toText(Lorg/apache/logging/log4j/core/layout/AbstractStringLayout$Serializer2;Lorg/apache/logging/log4j/core/LogEvent;Ljava/lang/StringBuilder;)Ljava/lang/StringBuilder; (PatternLayout.java:233)
  at org.apache.logging.log4j.core.layout.PatternLayout.encode(Lorg/apache/logging/log4j/core/LogEvent;Lorg/apache/logging/log4j/core/layout/ByteBufferDestination;)V (PatternLayout.java:218)
  at org.apache.logging.log4j.core.layout.PatternLayout.encode(Ljava/lang/Object;Lorg/apache/logging/log4j/core/layout/ByteBufferDestination;)V (PatternLayout.java:58)
  at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(Lorg/apache/logging/log4j/core/LogEvent;)V (AbstractOutputStreamAppender.java:177)
  at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(Lorg/apache/logging/log4j/core/LogEvent;)V (AbstractOutputStreamAppender.java:170)
  at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(Lorg/apache/logging/log4j/core/LogEvent;)V (AbstractOutputStreamAppender.java:161)
  at org.apache.logging.log4j.core.appender.RollingFileAppender.append(Lorg/apache/logging/log4j/core/LogEvent;)V (RollingFileAppender.java:308)
  at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:156)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:129)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:120)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:84)
  at org.apache.logging.log4j.core.appender.routing.RoutingAppender.append(Lorg/apache/logging/log4j/core/LogEvent;)V (RoutingAppender.java:227)
  at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:156)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:129)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:120)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:84)
  at org.apache.logging.log4j.core.appender.routing.RoutingAppender.append(Lorg/apache/logging/log4j/core/LogEvent;)V (RoutingAppender.java:227)
  at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:156)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:129)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:120)
  at org.apache.logging.log4j.core.config.AppenderControl.callAppender(Lorg/apache/logging/log4j/core/LogEvent;)V (AppenderControl.java:84)
  at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(Lorg/apache/logging/log4j/core/LogEvent;)V (LoggerConfig.java:448)
  at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(Lorg/apache/logging/log4j/core/LogEvent;)V (LoggerConfig.java:433)
  at org.apache.logging.log4j.core.config.LoggerConfig.log(Lorg/apache/logging/log4j/core/LogEvent;)V (LoggerConfig.java:417)
  at org.apache.logging.log4j.core.config.LoggerConfig.log(Ljava/lang/String;Ljava/lang/String;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (LoggerConfig.java:403)
  at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(Lorg/apache/logging/log4j/util/Supplier;Ljava/lang/String;Ljava/lang/String;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (AwaitCompletionReliabilityStrategy.java:63)
  at org.apache.logging.log4j.core.Logger.logMessage(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (Logger.java:146)
  at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (AbstractLogger.java:2163)
  at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (AbstractLogger.java:2118)
  at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Lorg/apache/logging/log4j/message/Message;Ljava/lang/Throwable;)V (AbstractLogger.java:2101)
  at org.apache.logging.log4j.spi.AbstractLogger.logMessage(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Ljava/lang/String;Ljava/lang/Throwable;)V (AbstractLogger.java:1995)
  at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(Ljava/lang/String;Lorg/apache/logging/log4j/Level;Lorg/apache/logging/log4j/Marker;Ljava/lang/String;Ljava/lang/Throwable;)V (AbstractLogger.java:1967)
  at org.apache.logging.slf4j.Log4jLogger.error(Ljava/lang/String;Ljava/lang/Throwable;)V (Log4jLogger.java:319)
  at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupOutput(Lorg/apache/pulsar/functions/instance/ContextImpl;)V (JavaInstanceRunnable.java:773)
  at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setup()V (JavaInstanceRunnable.java:212)
  at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run()V (JavaInstanceRunnable.java:234)
  at java.lang.Thread.run()V (Thread.java:748)`

And for the original error:
`Caused by: java.lang.NullPointerException: Cannot invoke "String.endsWith(String)" because "host" is null
	at org.jclouds.s3.filters.AwsHostNameUtils.parseServiceName(AwsHostNameUtils.java:164)
	at org.jclouds.s3.filters.Aws4SignerBase$ServiceAndRegion$AWSServiceAndRegion.<init>(Aws4SignerBase.java:97)
	at org.jclouds.s3.filters.Aws4SignerBase$ServiceAndRegion$AWSServiceAndRegion.<init>(Aws4SignerBase.java:93)
	at jdk.internal.reflect.GeneratedConstructorAccessor31.newInstance(Unknown Source)`

--- a/src/main/java/org/apache/pulsar/io/jcloud/BlobStoreAbstractConfig.java
+++ b/src/main/java/org/apache/pulsar/io/jcloud/BlobStoreAbstractConfig.java
@@ -23,6 +23,8 @@ import static com.google.common.base.Preconditions.checkNotNull;
 import static org.apache.commons.lang3.StringUtils.isNotBlank;
 import com.google.common.collect.ImmutableMap;
 import java.io.Serializable;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.time.Instant;
 import java.time.ZoneOffset;
 import java.time.format.DateTimeFormatter;
@@ -62,7 +64,6 @@ public class BlobStoreAbstractConfig implements Serializable {
 
     public static final String PROVIDER_AWSS3 = "aws-s3";
     public static final String PROVIDER_AWSS3V2 = "s3v2";
-    public static final String PROVIDER_S3 = "s3";
     public static final String PROVIDER_GCS = "google-cloud-storage";
 
     private String provider;
@@ -107,9 +108,12 @@ public class BlobStoreAbstractConfig implements Serializable {
     public void validate() {
         checkNotNull(provider, "provider not set.");
         checkNotNull(bucket, "bucket property not set.");
-        if (provider.equalsIgnoreCase(PROVIDER_AWSS3)) {
+        if (provider.equalsIgnoreCase(PROVIDER_AWSS3) || provider.equalsIgnoreCase(PROVIDER_AWSS3V2)) {
             checkArgument(isNotBlank(region) || isNotBlank(endpoint),
-                    "Either the aws-end-point or aws-region must be set");
+                    "Either the aws-end-point or aws-region must be set.");
+            if (isNotBlank(endpoint)) {
+                checkArgument(hasURIScheme(endpoint), "endpoint property needs to specify URI scheme.");
+            }
         }
 
         if (!formatMap.containsKey(StringUtils.lowerCase(formatType))) {
@@ -120,7 +124,7 @@ public class BlobStoreAbstractConfig implements Serializable {
                 && !partitionerType.equalsIgnoreCase("default")) {
             // `default` option is for backward compatibility
             throw new IllegalArgumentException(
-                    "partitionerType property not set properly, available options: partition / time");
+                    "partitionerType property not set properly, available options: partition / time.");
         }
         if (PartitionerType.TIME.name().equalsIgnoreCase(partitionerType)) {
             if (StringUtils.isNoneBlank(timePartitionPattern)) {
@@ -145,16 +149,25 @@ public class BlobStoreAbstractConfig implements Serializable {
 
         if ("bytes".equalsIgnoreCase(formatType)) {
             checkArgument(StringUtils.isNotEmpty(bytesFormatTypeSeparator),
-                    "bytesFormatTypeSeparator cannot be empty when formatType is 'bytes'");
+                    "bytesFormatTypeSeparator cannot be empty when formatType is 'bytes'.");
             checkArgument(StringUtils.startsWith(bytesFormatTypeSeparator, "0x"),
-                    "bytesFormatTypeSeparator should be a hex encoded string, which starts with '0x'");
+                    "bytesFormatTypeSeparator should be a hex encoded string, which starts with '0x'.");
         }
 
         if (pendingQueueSize <= 0) {
             pendingQueueSize = batchSize * 10;
         }
         checkArgument(pendingQueueSize > 0, "pendingQueueSize must be a positive integer.");
-        checkArgument(pendingQueueSize > batchSize, "pendingQueueSize must be large than batchSize");
+        checkArgument(pendingQueueSize > batchSize, "pendingQueueSize must be large than batchSize.");
+    }
+
+    private static boolean hasURIScheme(String endpoint) {
+        try {
+            URI uri = new URI(endpoint);
+            return isNotBlank(uri.getScheme());
+        } catch (URISyntaxException e) {
+            return false;
+        }
     }
 
 }

--- a/src/test/java/org/apache/pulsar/io/jcloud/ConnectorConfigTest.java
+++ b/src/test/java/org/apache/pulsar/io/jcloud/ConnectorConfigTest.java
@@ -42,7 +42,7 @@ public class ConnectorConfigTest {
         config.put("secretAccessKey", "aws-s3");
         config.put("bucket", "testbucket");
         config.put("region", "localhost");
-        config.put("endpoint", "us-standard");
+        config.put("endpoint", "https://us-standard");
         config.put("pathPrefix", "pulsar/");
         config.put("formatType", "avro");
         config.put("partitionerType", "default");
@@ -73,7 +73,7 @@ public class ConnectorConfigTest {
         config.put("secretAccessKey", "aws-s3");
         config.put("bucket", "testbucket");
         config.put("region", "localhost");
-        config.put("endpoint", "us-standard");
+        config.put("endpoint", "https://us-standard");
         config.put("formatType", "avro");
         config.put("partitionerType", "time");
         config.put("timePartitionPattern", "yyyy-MM-dd");
@@ -133,7 +133,7 @@ public class ConnectorConfigTest {
         config.put("secretAccessKey", "aws-s3");
         config.put("bucket", "testbucket");
         config.put("region", "localhost");
-        config.put("endpoint", "us-standard");
+        config.put("endpoint", "https://us-standard");
         config.put("formatType", "avro");
         config.put("partitionerType", "default");
         config.put("timePartitionPattern", "yyyy-MM-dd");
@@ -177,12 +177,32 @@ public class ConnectorConfigTest {
     }
 
     @Test
+    public final void throwOnMalformedEndpointTest() {
+        Map<String, Object> config = new HashMap<>();
+        config.put("provider", PROVIDER_AWSS3);
+        config.put("accessKeyId", "aws-s3");
+        config.put("secretAccessKey", "aws-s3");
+        config.put("bucket", "testbucket");
+        config.put("region", "us-east-1");
+        config.put("endpoint", "s3.us-east-1.amazonaws.com"); // no URI scheme
+        config.put("pathPrefix", "pulsar/");
+        config.put("formatType", "avro");
+        config.put("partitionerType", "default");
+        config.put("timePartitionPattern", "yyyy-MM-dd");
+        config.put("timePartitionDuration", "2d");
+        config.put("batchSize", 10);
+
+        Assert.assertThrows(IllegalArgumentException.class,
+                () -> CloudStorageSinkConfig.load(config).validate());
+    }
+
+    @Test
     public final void loadFromMapCredentialFromSecretTest() throws IOException {
         Map<String, Object> config = new HashMap<>();
         config.put("provider", "aws-s3");
         config.put("bucket", "testbucket");
         config.put("region", "localhost");
-        config.put("endpoint", "us-standard");
+        config.put("endpoint", "https://us-standard");
         config.put("formatType", "avro");
         config.put("partitionerType", "default");
         config.put("timePartitionPattern", "yyyy-MM-dd");
@@ -212,7 +232,7 @@ public class ConnectorConfigTest {
         config.put("secretAccessKey", "aws-s3");
         config.put("bucket", "testbucket");
         config.put("region", "localhost");
-        config.put("endpoint", "us-standard");
+        config.put("endpoint", "https://us-standard");
         config.put("pathPrefix", "pulsar/");
         config.put("formatType", "bytes");
         config.put("partitionerType", "default");

